src/transport/transporter.js

```diff
@@ -46,52 +44,63 @@ function Transporter(configuration, messageHandler, debugLogging, callback) {
-    var ipcTransport = new IpcTransport(ipcAddress, configuration.connectionTimeout, messageHandler, function (error) {
-      resultAggregator.ipcTransports[index] = (error !== null) ? null : ipcTransport;
-      // TODO: propagate the error up to the caller for reporting
-      checkIfComplete(this, resultAggregator, callback);
-    }.bind(this));
-  }.bind(this));
-  configuration.wsAddresses.forEach(function (wsAddress, index) {
-    var wsTransport = new WsTransport(wsAddress, configuration.connectionTimeout, messageHandler, function (error) {
-      resultAggregator.wsTransports[index] = (error !== null) ? null : wsTransport;
-      // TODO: propagate the error up to the caller for reporting
-      checkIfComplete(this, resultAggregator, callback);
-    }.bind(this));
-  }.bind(this));
-  configuration.httpAddresses.forEach(function (httpAddress, index) {
-    var httpTransport = new HttpTransport(httpAddress, configuration.connectionTimeout, messageHandler, function (error) {
-      resultAggregator.httpTransports[index] = (error !== null) ? null : httpTransport;
-      // TODO: propagate the error up to the caller for reporting
-      checkIfComplete(this, resultAggregator, callback);
+  // initiate connections to all provided addresses, only callback one that connects
+  someSeries([
+    function (callback) {
+      var web3Transport = new Web3Transport(messageHandler, function (error) {
+        if (error !== null) return callback(null);
+        return callback(web3Transport);
+      });
+    },
+    function (callback) {
+      if (configuration.ipcAddresses.length === 0) return callback(null);
+      someSeries(configuration.ipcAddresses,
+        function (ipcAddress, next) {
+          var ipcTransport = new IpcTransport(ipcAddress, configuration.connectionTimeout, messageHandler, function (error) {
+            if (error !== null) return next(null);
+            return next(ipcTransport);
+          });
+        },
+        callback);
+    },
+    function (callback) {
+      if (configuration.wsAddresses.length === 0) return callback(null);
+      someSeries(configuration.wsAddresses,
+        function (wsAddress, next) {
+          var wsTransport = new WsTransport(wsAddress, configuration.connectionTimeout, messageHandler, function (error) {
+            if (error !== null) return next(null);
+            return next(wsTransport);
+          });
+        },
+        callback);
+    },
+    function (callback) {
+      if (configuration.httpAddresses.length === 0) return callback(null);
+      someSeries(configuration.httpAddresses,
+        function (httpAddress, next) {
+          var httpTransport = new HttpTransport(httpAddress, configuration.connectionTimeout, messageHandler, function (error) {
+            if (error !== null) return next(null);
+            return next(httpTransport);
+          });
+        },
+        callback);
+    }],
+    function (callback, next) {
```

Comment:
The use of `callback` in this function is confusing.  `callback` is actually the item in the array that you're iterating over, which is itself passed a parameter that (in the array) is also referred to as `callback`.

Suggestion: use `function (tryTransportType, nextTransportType)` instead.